### PR TITLE
Support Secp256r1

### PR DESCRIPTION
**Pull request by mitschabaude**
closes https://github.com/o1-labs/o1js/issues/1724

this adds ECDSA (and general) support for the curve **secp256r1**, also known as **P-256**, which is the standard curve used in ECDSA outside the crypto world, due to its recommendation by NIST; see [P-256 in this document](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-186.pdf)

Most changes are in https://github.com/o1-labs/o1js-bindings/pull/310

changes here:
* properly treat and pass on curve parameter `a` in EC addition and doubling methods
  * no circuit changes! the parameter was just missing in witness generation
* add secp256r1 to the curve and ECDSA unit tests

## Efficiency

Sadly, secp256r1 doesn't support the endomorphism we have implemented, because that endomorphism crucially relies on the `a = 0` property. Therefore, in our ECDSA circuit the GLV method is not used, which means significant constraint savings get lost:
* ECDSA on secp256k1: **32k** constraints
* ECDSA on secp256r1: **40k** constraints

### CODE REVIEW

**Comment by mitschabaude** [comment]
@Trivo25 can you get me the necessary reviewers for this?

**Review by Geometer1729 — COMMENTED** [review]
LGTM

**Comment by mitschabaude** [comment]
Thanks for the review @Geometer1729! I also need one [here](https://github.com/o1-labs/o1js-bindings/pull/310) and then I need you to merge both PRs since I don't have permission to

**Comment by mitschabaude** [comment]
Oh and @Trivo25 I also need admin review for changing the bindings commit

**Comment by Geometer1729** [comment]
@mitschabaude Can you run `npm run build:update-bindings` and push the changes that creates in `o1js-bindings`? I think the bindings not having been recompiled may be why CI is failing.

**Comment by mitschabaude** [comment]
> @mitschabaude Can you run `npm run build:update-bindings` and push the changes that creates in `o1js-bindings`? I think the bindings not having been recompiled may be why CI is failing.

no @Geometer1729, I didn't change anything that needs compiling. also, CI isn't failing, just the benchmark job is, and that one _always_ fails on forks

**Comment by mitschabaude** [comment]
updating the bindings is just needed when you need to recompile ocaml or rust sources

**Comment by Geometer1729** [comment]
Oh, I thought it was also affected by some of the content of o1js-bindings. Fair enough sorry about that.

File: src/lib/provable/test/ecdsa.unit-test.ts
Comment (45930) [inline src/lib/provable/test/ecdsa.unit-test.ts:222]: Could we add some test vectors to verify this implementation?

Comment (mitschabaude) [reply src/lib/provable/test/ecdsa.unit-test.ts:222]: yes! I'll do that

**Comment by 45930** [comment]
Amazing work!  Florian has final approval, but it LGTM

**Comment by mitschabaude** [comment]
the test is failing because we're still running node 18 in CI, where the web crypto API is missing 😬 

do you mind if I change it to node 20?

**Comment by mitschabaude** [comment]
need workflow approval again 😢 @Trivo25

**Comment by mitschabaude** [comment]
and again @Trivo25!

**Comment by mitschabaude** [comment]
@nicc @mrmr1993 that's why I need write access to the repo ^^ https://github.com/o1-labs/o1js/pull/1885#issuecomment-2436878827

**Comment by Trivo25** [comment]
@hattyhattington17 can we get your eyes on this too?

**Comment by Trivo25** [comment]
@mitschabaude fyi I will get this merged before the next release

**Review by Trivo25 — APPROVED** [review]

File: src/lib/provable/gadgets/elliptic-curve.ts
Comment (querolita) [inline src/lib/provable/gadgets/elliptic-curve.ts:131]: In [PR310](https://github.com/zksecurity/o1js-bindings/blob/e081466d5e435fe3fbd32fa83ef5c0ef3f030dec/crypto/elliptic-curve.ts#L578) this `m` is computed with a single modular operation, meaning: `let m = mod((3n * x * x + a) * d, p);`. Cannot we remove one `mod` from this line?

**Review by querolita — APPROVED** [review]

**Review by hattyhattington17 — APPROVED** [review]

File: src/lib/provable/gadgets/elliptic-curve.ts
Comment (mitschabaude) [reply src/lib/provable/gadgets/elliptic-curve.ts:131]: good question! I think using `mod()` twice is better actually, because both x and d are large field elements. so x^2 is already a bigint of 2x the length, and it's more efficient to first reduce it (with `mod()`) back to 1x the length before doing another bigint multiplication with d. In the code you mention, we get an intermediate result of 2x * 1x = 3x the length

**Pull request merged**
